README.md
# `AnnotationHub` for proteomics

<!-- README.md is generated from README.Rmd. Please edit that file -->



The aim of this package is to offer access to mass spectrometry and
proteomics data throught the *[AnnotationHub](http://bioconductor.org/packages/AnnotationHub)*
infrastructure.




```r
library("AnnotationHub")
#> Loading required package: BiocGenerics
#> Loading required package: methods
#> Loading required package: parallel
#> 
#> Attaching package: 'BiocGenerics'
#> 
#> The following objects are masked from 'package:parallel':
#> 
#>     clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
#>     clusterExport, clusterMap, parApply, parCapply, parLapply,
#>     parLapplyLB, parRapply, parSapply, parSapplyLB
#> 
#> The following objects are masked from 'package:stats':
#> 
#>     IQR, mad, xtabs
#> 
#> The following objects are masked from 'package:base':
#> 
#>     anyDuplicated, append, as.data.frame, as.vector, cbind,
#>     colnames, do.call, duplicated, eval, evalq, Filter, Find, get,
#>     grep, grepl, intersect, is.unsorted, lapply, lengths, Map,
#>     mapply, match, mget, order, paste, pmax, pmax.int, pmin,
#>     pmin.int, Position, rank, rbind, Reduce, rownames, sapply,
#>     setdiff, sort, table, tapply, union, unique, unlist, unsplit
ah <- AnnotationHub()
#> snapshotDate(): 2015-11-19
ah
#> AnnotationHub with 35372 records
#> # snapshotDate(): 2015-11-19 
#> # $dataprovider: BroadInstitute, UCSC, Ensembl, ftp://ftp.ncbi.nlm.nih....
#> # $species: Homo sapiens, Mus musculus, Bos taurus, Pan troglodytes, Da...
#> # $rdataclass: GRanges, BigWigFile, FaFile, ChainFile, OrgDb, Inparanoi...
#> # additional mcols(): taxonomyid, genome, description, tags,
#> #   sourceurl, sourcetype 
#> # retrieve records with, e.g., 'object[["AH2"]]' 
#> 
#>             title                                               
#>   AH2     | Ailuropoda_melanoleuca.ailMel1.69.dna.toplevel.fa   
#>   AH3     | Ailuropoda_melanoleuca.ailMel1.69.dna_rm.toplevel.fa
#>   AH4     | Ailuropoda_melanoleuca.ailMel1.69.dna_sm.toplevel.fa
#>   AH5     | Ailuropoda_melanoleuca.ailMel1.69.ncrna.fa          
#>   AH6     | Ailuropoda_melanoleuca.ailMel1.69.pep.all.fa        
#>   ...       ...                                                 
#>   AH49587 | org.Ce.eg.db.sqlite                                 
#>   AH49588 | org.Xl.eg.db.sqlite                                 
#>   AH49589 | org.Sc.sgd.db.sqlite                                
#>   AH49590 | org.Dr.eg.db.sqlite                                 
#>   AH49591 | org.Pf.plasmo.db.sqlite
```

Let's start by querying the entries that originate from the PRIDE
database:


```r
query(ah, "PRIDE")
#> AnnotationHub with 4 records
#> # snapshotDate(): 2015-11-19 
#> # $dataprovider: PRIDE
#> # $species: Erwinia carotovora
#> # $rdataclass: AAStringSet, MSnSet, mzRident, mzRpwiz
#> # additional mcols(): taxonomyid, genome, description, tags,
#> #   sourceurl, sourcetype 
#> # retrieve records with, e.g., 'object[["AH49006"]]' 
#> 
#>             title                                                         
#>   AH49006 | PXD000001: Erwinia carotovora and spiked-in protein fasta file
#>   AH49007 | PXD000001: Peptide-level quantitation data                    
#>   AH49008 | PXD000001: raw mass spectrometry data                         
#>   AH49009 | PXD000001: MS-GF+ identiciation data
```

Or those of a specific project


```r
ah[grep("PXD000001", ah$title)]
#> AnnotationHub with 4 records
#> # snapshotDate(): 2015-11-19 
#> # $dataprovider: PRIDE
#> # $species: Erwinia carotovora
#> # $rdataclass: AAStringSet, MSnSet, mzRident, mzRpwiz
#> # additional mcols(): taxonomyid, genome, description, tags,
#> #   sourceurl, sourcetype 
#> # retrieve records with, e.g., 'object[["AH49006"]]' 
#> 
#>             title                                                         
#>   AH49006 | PXD000001: Erwinia carotovora and spiked-in protein fasta file
#>   AH49007 | PXD000001: Peptide-level quantitation data                    
#>   AH49008 | PXD000001: raw mass spectrometry data                         
#>   AH49009 | PXD000001: MS-GF+ identiciation data
```

To see the metadata of a specific entry, we use its AnnotationHub
entry number inside single `[`


```r
ah["AH49008"]
#> AnnotationHub with 1 record
#> # snapshotDate(): 2015-11-19 
#> # names(): AH49008
#> # $dataprovider: PRIDE
#> # $species: Erwinia carotovora
#> # $rdataclass: mzRpwiz
#> # $title: PXD000001: raw mass spectrometry data
#> # $description: Four human TMT spliked-in proteins in an Erwinia caroto...
#> # $taxonomyid: 554
#> # $genome: NA
#> # $sourcetype: mzML
#> # $sourceurl: ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2012/03/PXD0...
#> # $sourcelastmodifieddate: NA
#> # $sourcesize: NA
#> # $tags: Proteomics, TMT6, LTQ Orbitrap Velos, PMID:23692960 
#> # retrieve record with 'object[["AH49008"]]'
```

To access the actual data, raw mass spectrometry data in this case, we
double the `[[`


```r
library("mzR")
#> Loading required package: Rcpp
rw <- ah[["AH49008"]]
#> loading from cache '/home/lg390/.AnnotationHub/55314'
rw
#> Mass Spectrometry file handle.
#> Filename:  55314 
#> Number of scans:  7534
```

In this case, we have an instance of class mzRpwiz,
that can be processed as anticipated.

In the short demonstration above, we had **direct** and
**standardised** access to the raw data, without a need to manually
open this raw data or worry about the file format. The data was
prepared and converted into a **standard Bioconductor data types** for
immediate consumption by the user. This is also valid for other
relevant data types such as identification results, fasta files or
protein of peptide quantitation data.

See the `ProteomicsAnnotationHubData` vignette for details on
available data and how to add new proteomics data to AnnotationHub.


NEWS
CHANGES IN VERSION 1.11.1
-------------------------
- Replace BiocInstaller::biocVersion with BiocManager::version <2018-09-03 Mon>

CHANGES IN VERSION 1.11.0
-------------------------
- New version for Bioc devel
- Import BiocManager

CHANGES IN VERSION 1.3.3
------------------------
 o Fix formatting in vignette <2016-06-26 Sun>

CHANGES IN VERSION 1.1.2
------------------------
 o Major rewrite of the data preparation, now relying on simple dcf
   files as input and intermediate PAHD objects. <2015-12-17 Thu>
 o import read.table <2016-03-29 Tue>
 o replace curl code by RCurl::getURL <2016-03-29 Tue>

CHANGES IN VERSION 1.1.1
------------------------
 o Manually setting flInfo in .ftpfileinfo2 for files located on S3
   <2015-12-15 Tue>

CHANGES IN VERSION 1.1.0
------------------------
 o Bioc devel 3.3

CHANGES IN VERSION 1.0.0
------------------------
 o Bioc release 3.2

CHANGES IN VERSION 0.99.3
-------------------------
 o removing .get1 methods and pointing to relevant AnnotationHub file
   <2015-10-09 Fri>

CHANGES IN VERSION 0.99.2
-------------------------
 o Use AnnotationHub::query() instead of manual subsetting
   <2015-09-02 Wed>

CHANGES IN VERSION 0.99.1
-------------------------
 o Submission of Bioc (Issue 1292) <2015-08-25 Tue>

CHANGES IN VERSION 0.99.0
-------------------------
 o  First release with data PXD000001


R/makePXD000001.R
.prideDir <- "pride/data/archive/2012/03/PXD000001/"
.prideFullUrl <- file.path(.prideBaseUrl, .prideDir)

fls <- c("TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML",
         "F063721.dat-mztab.txt",
         "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid",
         "erwinia_carotovora.fasta")

preparers <- c("PXD000001MzMLToMzRPwizPreparer",
               "PXD000001MzTabToMSnSetPreparer",
               "PXD000001MzidToMzRidentPreparer",
               "PXD000001MzMLToAAStringSetPreparer")

.locationPrefix <- c(.prideBaseUrl,  ## file in PRIDE ftp
                     .amazonBaseUrl, ## file on AH S3
                     .prideBaseUrl,
                     .prideBaseUrl)

.SourceType <- c("mzML", "mzTab", "mzid", "FASTA")
.DispatchClass <- .RDataClass <- c("mzRpwiz", "MSnSet", "mzRident", "AAStringSet")

## Only required if the rda file is prepared on the AH Amazon instance
PXD00001MzTabToMSnSet <- function(ahm) {
    ## Imports: MSnbase
    if (file.exists(outputFile(ahm))) 
        return(outputFile(ahm))

    fin <- inputFiles(ahm) ## get file name on the ftp site
    fout <- outputFile(ahm)    
    fl <- download.file(fin, fout)
    msnset <- MSnbase::readMzTabData(fin, what = "PEP")
    save(msnset, file = "F063721.dat-MSnSet.rda")
   
    outputFile(ahm)    
}

.Recipe <- c(NA_character_,
             "PXD000001MzTabToMSnSet",
             NA_character_,
             NA_character_)

flInfo <- Map(.ftpFileInfo, url = .prideFullUrl,
              filename = fls, tag = NA_character_)

flInfo <- do.call(rbind, flInfo)
flInfo$date <- as.character(flInfo$date)

.Title <- "Four human TMT spliked-in proteins in an Erwinia carotovora background"
.Description <- .expandLine("Expected reporter ion ratios: Erwinia peptides: 1:1:1:1:1:1;
                            Enolase spike (sp|P00924|ENO1_YEAST): 10:5:2.5:1:2.5:10;
                            BSA spike (sp|P02769|ALBU_BOVIN): 1:2.5:5:10:5:1;
                            PhosB spike (sp|P00489|PYGM_RABIT): 2:2:2:2:1:1;
                            Cytochrome C spike (sp|P62894|CYC_BOVIN): 1:1:1:1:1:2.")

.Genome <- NA_character_
.TaxonomyId <- 554L
.Species <- "Erwinia carotovora"
.DataProvider <- ProteomicsAnnotationHubDataProviders['PRIDE']
.Maintainer <- "Laurent Gatto <lg390@cam.ac.uk>"
.RDataPath <- file.path(.prideDir, fls)
.Tags <- c("Proteomics", "TMT6", "LTQ Orbitrap Velos", "PMID:23692960")



for (i in seq_along(fls)) {
    AHMRecord <- 
        AnnotationHubMetadata(
            Title = .Title,
            Description = .Description,
            SourceType = .SourceType[i],
            SourceUrl = flInfo[i, 1],
            SourceVersion = flInfo[i, 2],
            Species = .Species,
            Genome = .Genome,
            TaxonomyId = .TaxonomyId,
            RDataPath = .RDataPath[i],
            DataProvider = .DataProvider,
            Maintainer = .Maintainer,
            RDataClass =  .RDataClass[i],
            BiocVersion  =  BiocInstaller::biocVersion(),
            RDataDateAdded = Sys.time(),
            Location_Prefix = .locationPrefix[i],
            Recipe = .Recipe[i],
            DispatchClass = .DispatchClass,
            Tags = .Tags)
    makeAnnotationHubResource(preparers[i],
                              function(currentMetadata, justRunUnitTest) 
                                  AHMRecord)
}
